### ISSUE TITLE
Package camlp4 has a bad versioning system

### ISSUE THREAD

**Issue by yannl35133** [ocaml/opam#4762]
The algorithm to detect which version of camlp4 is incompatible with the way it is currently named :

Version `camlp4.4.11+system` is considered more up-to-date than `camlp4.4.11+1`, even if they are supposed to be the same version.

I don't know how to solve this issue best; I believe a system of virtual package with two dependencies according to whether the compiler is opam-installed or system wide could work, but I have no experience in making this work.

See [this bug](https://github.com/camlp4/camlp4/issues/157) in `camlp4/camlp4`

**Comment by kit-ty-kate** [comment]
This issue doesn't seem to belong to opam-repository but rather what you are saying is that opam upgrade itself is confusing.

I've move the issue to the opam bugtracker and changed the title to be more accurate

**Comment by dra27** [comment]
There's a definite bug:

```
  - ocaml.4.13.0
    ∗ camlp4.4.12+system is installed and requires ocaml (>= 4.12 & < 4.13)
```
that should not be referring to `camlp4.4.12+system`, as presumably it's `camlp4.4.12+1` which is installed.

**Comment by dra27** [comment]
The other part is a (good) feature request - the `ocaml` package is `{post}` dependency of `ocaml-base-compiler`. We should not display upgrade messages about packages in the invariant - so the `ocaml` package should not be considered for upgrade because it's a `{post}` dependency of the switch invariant and the upgrade would require relaxing the switch invariant.

**Comment by yannl35133** [comment]
The original bug relating to `camlp4`'s versioning remains : opam sees `camlp4.4.12+system` as an upgrade to `camlp4.4.12+1` when it is not.

**Comment by dra27** [comment]
That part is not a bug - `4.12+system` is greater than `4.12+1` and so is an upgrade.

**Comment by yannl35133** [comment]
Wrong wording on my part : it is a *problem* of the versioning (from camlp4 directly), not a *bug* in opam

**Comment by dra27** [comment]
Yes, indeed 🙂 For camlp4, it would be better to have `camlp4` as a virtual package with `camlp4-base` and `camlp4-system` (cf. `ocaml` with `ocaml-base-compiler` and `ocaml-system`). However, the `ocaml` package has a special bit of support to detect the presence and version of the installed system OCaml compiler, and that's not at the moment generalisable to `camlp4`.
